controllers/imagebasedgroupupgrade_test.go

```diff
@@ -0,0 +1,336 @@
+					State: "complete",
+				},
+				{
+					Name:  "spoke4",
+					State: "complete",
+				},
+				{
+					Name:          "spoke6",
+					State:         "InProgress",
+					CurrentAction: &namePrep,
+				},
+			},
+		},
+	}
+	ibgu := &ibguv1alpha1.ImageBasedGroupUpgrade{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "name",
+			Namespace: "namespace",
+		},
+		Spec: ibguv1alpha1.ImageBasedGroupUpgradeSpec{
+
+			RolloutStrategy: ibguv1alpha1.RolloutStrategy{
+				Timeout:        50,
+				MaxConcurrency: 2,
+			},
+			ClusterLabelSelectors: []v1.LabelSelector{
+
+				{
+					MatchLabels: map[string]string{
+						"common": "true",
+					},
+				},
+			},
+			IBUSpec: lcav1alpha1.ImageBasedUpgradeSpec{
+				SeedImageRef: lcav1alpha1.SeedImageRef{
+					Version: "version",
+					Image:   "image",
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		objs := []client.Object{}
+		fakeClient, err := getFakeClientFromObjects(objs...)
+		for _, cgu := range test.CGUs {
+			err := fakeClient.Create(context.TODO(), &cgu)
+			if err != nil {
+				panic(err)
+			}
+		}
+		if err != nil {
+			t.Errorf("error in creating fake client")
+		}
+		reconciler := IBGUReconciler{Client: fakeClient, Scheme: testscheme, Log: logr.Discard()}
+		err = reconciler.syncStatusWithCGUs(context.Background(), ibgu)
+		assert.NoError(t, err)
+		assert.ElementsMatch(t, test.expectedClustersStatus, ibgu.Status.Clusters)
+	}
+}
+
+func TestEnsureManifests(t *testing.T) {
+	tests := []struct {
+		name         string
+		actions      []ibguv1alpha1.ImageBasedUpgradeAction
+		expectedMWRS []string
+		CGUs         []v1alpha1.ClusterGroupUpgrade
+		expectedCGUs []v1alpha1.ClusterGroupUpgrade
+	}{
+		{
+			name: "without append",
+			actions: []ibguv1alpha1.ImageBasedUpgradeAction{
+				"Prep",
+				"Upgrade",
+				"Finalize",
+			},
+			expectedMWRS: []string{"name-prep", "name-upgrade", "name-finalize"},
+			expectedCGUs: []v1alpha1.ClusterGroupUpgrade{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name: "name-prep-upgrade-finalize",
+					},
+				},
+			},
+		},
+		{
+			name: "append",
+			actions: []ibguv1alpha1.ImageBasedUpgradeAction{
+				"Prep",
+				"Upgrade",
+				"Finalize",
+			},
+			expectedMWRS: []string{"name-prep", "name-upgrade", "name-finalize"},
+			CGUs: []v1alpha1.ClusterGroupUpgrade{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "name-prep",
+						Namespace: "namespace",
+						Labels: map[string]string{
+							utils.CGUOwnerIBGULabel: "name",
+						},
+					},
+					Spec: v1alpha1.ClusterGroupUpgradeSpec{
+						ManifestWorkTemplates: []string{
+							"name-prep",
+						},
+					},
+				},
+			},
+			expectedCGUs: []v1alpha1.ClusterGroupUpgrade{
+				{
+					ObjectMeta: v1.ObjectMeta{
+						Name:      "name-prep",
+						Namespace: "namespace",
+					},
+					Spec: v1alpha1.ClusterGroupUpgradeSpec{
+						ManifestWorkTemplates: []string{
+							"name-prep",
+						},
+					},
+				},
+				{
+					ObjectMeta: v1.ObjectMeta{
```

Comment:
What happens if the append is done before the first CGU completes?

Reply:
The next CGU is created but it has the previous CGU as the blocking CGU.

Reply:
Great idea. Unfortunately, the blocking feature was never practically usable at scale. The previous CGU has to be 100% successful for the next one to start.